src/System/Configuration/DataSets/ConfigurationDataSetLoader.cs

```diff
@@ -275,19 +283,30 @@ private CachedConfigurationDataSet<T> DataSets
 		/// </summary>
 		/// <typeparam name="TDataSet">The data set type.</typeparam>
 		private sealed class CachedConfigurationDataSet<TDataSet>
-			where TDataSet : class, IConfigurationDataSet, new()
+			where TDataSet : class, IConfigurationDataSet
 		{
 			/// <summary>
 			/// The default dataset for creating a data set instance.
 			/// </summary>
 			private readonly TDataSet m_dataSetDefault;
 
 
+			/// <summary>
+			/// Factory used to create a new data set.
+			/// </summary>
+			private readonly IDataSetFactory<TDataSet> m_dataSetFactory;
+
+
 			/// <summary>
 			/// Initializes a new instance of the <see cref="CachedConfigurationDataSet{TDataSet}" /> class.
 			/// </summary>
 			/// <param name="dataSetDefault">The data set default.</param>
-			public CachedConfigurationDataSet(TDataSet dataSetDefault) => m_dataSetDefault = dataSetDefault;
+			/// <param name="dataSetFactory">Data set factory.</param>
+			public CachedConfigurationDataSet(TDataSet dataSetDefault, IDataSetFactory<TDataSet> dataSetFactory)
+			{
+				m_dataSetDefault = dataSetDefault;
+				m_dataSetFactory = Code.ExpectsArgument(dataSetFactory, nameof(dataSetFactory), TaggingUtilities.ReserveTag(0));
```

Comment:
Would it make sense to make dataSetFactory optional here too? Especially if m_dataSetDefault is set

Reply:
The risk of the optional parameter is that it could hide potential errors in the future, in this case default values would be used and in most cases the gates won't loaded. 

I even think that as a next step we should remove the default values of OmexTip.xml and OmexGates.xml, I think it is better that they are specified explicitly.

What do you think?

---
In reply to: [306892478](https://github.com/microsoft/Omex/pull/66#discussion_r306892478) [](ancestors = 306892478)

Reply:
Discussed offline. We will keep it like that from maintainability perspective.

---
In reply to: [307202321](https://github.com/microsoft/Omex/pull/66#discussion_r307202321) [](ancestors = 307202321,306892478)